### PR TITLE
fix: Post selector with Go templates in ApplicationSet

### LINKED ISSUE #12524: Post selectors broken when using goTemplate: true
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

When goTemplate is enabled, the [post selectors](https://argo-cd.readthedocs.io/en/stable/user-guide/application-set/#post-selector-all-generators) do not work anymore.

I assume this problem is caused by the params needing to be prefixed with a dot when goTemplate is active, unfortunately prefixing a selectorKey is impossible as this will fail [during creation](https://github.com/argoproj/argo-cd/blob/master/applicationset/generators/generator_spec_processor.go#L30)  in the library's [validation for the requirements](https://github.com/kubernetes/apimachinery/blob/bfd2aff97e594f6aad77acbe2cbbe190acc93cbc/pkg/util/validation/validation.go#L43)

[This issue will be the exact same problem](https://github.com/argoproj/argo-cd/issues/11664), seeing that goTemplate in this case is also enabled. The assumption that keys with dots are completely unsupported is incorrect, when goTemplate is disabled keys with dots work perfectly fine (just not prefixed).

**To Reproduce**

Simply create an applicationset that uses both goTemplate: true and post selectors, neither version (with or without prefix) will work

**Version**

ArgoCD v2.6.2

### PR DESCRIPTION
**Pull request by m13t**
This PR fixes #12524 by allow users to reference nested properties when using Go template syntax in application sets. Previously only top-level string keys could be used as post selectors. When matching labels, this PR will now allow referencing nested keys by concatenating each key with a dot (`.`).

The original match values tests have been duplicated with Go templates enabled to provide more extensive tests.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

**Comment by m13t** [comment]
I've made a slight addition to my original commit for this to create a generic map flattening function as this would also be useful for supported nested keys in the merge generator. Assuming this PR is okay I'll look to tackle that as a separate PR.

File: applicationset/generators/generator_spec_processor.go [hunk 1]
```diff
@@ -71,8 +71,8 @@ func Transform(requestedGenerator argoprojiov1alpha1.ApplicationSetGenerator, al
 		}
 		var filterParams []map[string]interface{}
 		for _, param := range params {
-
-			if requestedGenerator.Selector != nil && !selector.Matches(labels.Set(keepOnlyStringValues(param))) {
+			flatParam := flattenParameters(param)
```
Comment (crenshaw-dev) [inline applicationset/generators/generator_spec_processor.go:74]: We're using https://github.com/jeremywohl/flatten elsewhere. Is it not a good fit for this use case?

File: applicationset/generators/generator_spec_processor_test.go [hunk 2]
```diff
@@ -73,7 +73,82 @@ func TestMatchValues(t *testing.T) {
...
+				emptyTemplate(),
+				&applicationSetInfo, nil)
+
+			assert.NoError(t, err)
+			assert.ElementsMatch(t, testCase.expected, results[0].Params)
+		})
+	}
+}
+
+func TestMatchValuesGoTemplate(t *testing.T) {
```
Comment (crenshaw-dev) [inline applicationset/generators/generator_spec_processor_test.go:97]: Would you mind adding a test for list values, just so it's documented somewhere? Probably an edge case, but good to cover.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

**Comment by crenshaw-dev** [comment]
Thanks so much for this PR! And yes, fixing `mergeKeys` would also be huge.

File: applicationset/generators/generator_spec_processor.go (hunk 1 above)
Comment (m13t) [reply applicationset/generators/generator_spec_processor.go:74]: I’ll look in to this.

File: applicationset/generators/generator_spec_processor_test.go (hunk 2 above)
Comment (m13t) [reply applicationset/generators/generator_spec_processor_test.go:97]: Could you elaborate on this? Do you mean tests for values that slice types?

File: applicationset/generators/generator_spec_processor.go (hunk 1 above)
Comment (m13t) [reply applicationset/generators/generator_spec_processor.go:74]: So would this be acceptable?

```go
flatParam, err := flatten.Flatten(param, "", flatten.DotStyle)
if err != nil {
	log.WithError(err).WithField("generator", g).
		Error("error flattening params")
	if firstError == nil {
		firstError = err
	}
	continue
}

strParam := make(map[string]string, len(flatParam))
for k, v := range flatParam {
	strParam[k] = fmt.Sprintf("%v", v)
}

if requestedGenerator.Selector != nil && !selector.Matches(labels.Set(strParam)) {
	continue
}
filterParams = append(filterParams, param)
```

The Sprintf call to convert the interface value into a string feels a little janky but I guess most non-scalar values are handled by the flatten call and there's no real way for a user to put a complex type in there anyway.

Comment (crenshaw-dev) [reply applicationset/generators/generator_spec_processor.go:74]: Agreed, that has a janky feel to it. Maybe wrap the conversion and also use it in the other place we're calling `flatten.Flatten`? They both have to convert to string.

File: applicationset/generators/generator_spec_processor_test.go (hunk 2 above)
Comment (crenshaw-dev) [reply applicationset/generators/generator_spec_processor_test.go:97]: Yep, exactly. So if my generator returns `{"list": ["1", "2", "3"]}`, can I select on the second item in `list`? If so, what's the selector look like?

Comment (m13t) [reply applicationset/generators/generator_spec_processor_test.go:97]: So I've just done a quick test with this. For go templates It works fine but it seems without the go templates the tests fail as there's an expectation that the values will always be of type `map[string]interface{}` and then subsequently that the value of that interface is a string. I suspect this may be a greater piece of work to add that support in to the older style templates.

File: applicationset/generators/generator_spec_processor.go (hunk 1 above)
Comment (m13t) [reply applicationset/generators/generator_spec_processor.go:74]: Interestingly, the git generator actually uses the same sprintf conversion which is the only other occurrence of the Flatten function being used. I'll extract that out and update both references.

File: applicationset/generators/generator_spec_processor_test.go (hunk 2 above)
Comment (crenshaw-dev) [reply applicationset/generators/generator_spec_processor_test.go:97]: Ah interesting. No sweat then, I think it's a far-edge case.

Comment (m13t) [reply applicationset/generators/generator_spec_processor_test.go:97]: Okay, I've added a test case for the go templates but omitted the test for argo templates.

File: applicationset/generators/generator_spec_processor.go (hunk 1 above)
Comment (m13t) [reply applicationset/generators/generator_spec_processor.go]: Updated and removed the custom flatten implementation in place of the flatten library. I've wrapped it in a function to handle to type conversion from `map[string]interface{}` to `map[string]string` to clean things up.
However, the git generator wants a `map[string]interface{}` so no need for the conversion here so I have left that as-is.

**Comment by m13t** [comment]
@crenshaw-dev I've rebased this PR and there looks to be some issues with the recently merged in test changes you did on the matrix generator as part of https://github.com/m13t/argo-cd/commit/fafee48b3c208c6f5f9de89cd4cb83caf87d4c01.

Looks like a go templates is being passed to the test but it's not using Go templates so subsequently failing. Any ideas?

**Review by crenshaw-dev — APPROVED** [review]
Beautiful, thanks @m13t!

**Comment by crenshaw-dev** [comment]
@m13t would you mind opening PRs against release-2.5, release-2.6, and release-2.7? Cherry-pick bot is finicky.

**Comment by m13t** [comment]
@crenshaw-dev manual cherry-picked PRs are in
